### PR TITLE
dev(clickhouse): strip out comments before executing sql

### PR DESCRIPTION
**Pull request by hazzadous**
This is so we can easily copy/paste from e.g. Metabase by querying the
system.query_log. In metabase is doesn't display new lines (although you
can download to file for this), but it's not very practical.

### CODE REVIEW

**Comment by neilkakkar** [comment]
You could also use block comments inside queries instead? `/* */`

**Comment by hazzadous** [comment]
> You could also use block comments inside queries instead? `/* */`

Yes! Although I believe there is some concern about performance as well with including comments. @macobo what do you think?

**Comment by Twixes** [comment]
Block comments _are_ occasionally useful for debugging query builder mistakes IMO.

**Comment by hazzadous** [comment]
> > You could also use block comments inside queries instead? `/* */`
> 
> Yes! Although I believe there is some concern about performance as well with including comments. @macobo what do you think?

Although it will be a surprise for anyone coming to the code base to not be able to use line comments #devexperience

**Comment by macobo** [comment]
Not sure if tests broke or not.

> Yes! Although I believe there is some concern about performance as well with including comments. @macobo what do you think?

The performance problem is if comments end up repeated the SQL can surpass what clickhouse will accept.

> Block comments are occasionally useful for debugging query builder mistakes IMO.

You should still be able to backtrace what query this was from the sql itself. Including comments in production currently just causes trouble.

File: ee/clickhouse/client.py [hunk 1]
```diff
@@ -137,11 +137,23 @@ def sync_execute(query, args=None, settings=None, with_column_types=False):
                 print()
                 print(format_sql(query, args))
```
Comment (macobo) [inline ee/clickhouse/client.py:138]: This is doing another substitution and formatting on its own. It's worth refactoring this and avoid needing to do formatting N times in one call.

File: ee/clickhouse/test/test_client.py [hunk 2]
```diff
@@ -43,3 +45,23 @@ def test_cache_eviction(self):
...
+        See https://github.com/metabase/metabase/issues/14253
+
+        Note I'm not really testing much complexity, I trust that those will
+        come out as failures in other tests.
+        """
+        query = f"""
+            -- this request returns 1
+            SELECT 1
+        """
+        with patch.object(Client, "execute") as mock_execute:
```
Comment (macobo) [inline ee/clickhouse/test/test_client.py:64]: There already exists `capture_select_queries` helper. Use that instead.

File: ee/clickhouse/client.py [hunk 3]
```diff
@@ -133,26 +134,36 @@ def sync_execute(query, args=None, settings=None, with_column_types=False):
...
+            # because we want to strip the comments to make it easier to copy
+            # and past queries from the `system.query_log` easily with metabase
+            # (metabase doesn't show new lines, so with comments, you can't get
+            # a working query without exporting to csv or similar), we need to
+            # do it manually.
+            #
+            # We only want to try to substitue for SELECT queries, which
+            # clickhouse_driver at this moment in time decides based on the
+            # below predicate.
+            if isinstance(args, (list, tuple, types.GeneratorType)):
```
Comment (macobo) [inline ee/clickhouse/client.py:148]: ? When is this the case? This doesn't seem sane.

File: ee/clickhouse/client.py [hunk 4]
```diff
@@ -133,26 +134,36 @@ def sync_execute(query, args=None, settings=None, with_column_types=False):
...
+                rendered_sql = query
+            else:
+                rendered_sql = client.substitute_params(query, args or {})
+            formatted_sql = sqlparse.format(rendered_sql, strip_comments=True)
+
+            # Adds in a /* */ so we can look in clickhouses `system.query_log`
+            # to easily marry up to the generating code.
+            annotated_sql, tags = _annotate_tagged_query(formatted_sql, args)
```
Comment (macobo) [inline ee/clickhouse/client.py:156]: Does this code block with ~20 lines of code belong in this function? When I was reading this function I'd not care about it.

File: ee/clickhouse/client.py [hunk 5]
```diff
@@ -207,24 +218,20 @@ def _notify_of_slow_query_failure(tags: Dict[str, Any]):
     incr("clickhouse_sync_execution_failure", tags=tags)
 
 
-def format_sql(sql, params, colorize=True):
-    substitute_params = (
-        ch_client.substitute_params if isinstance(ch_client, SyncClient) else ch_client._client.substitute_params
```
Comment (macobo) [inline ee/clickhouse/client.py:200]: Was this if statement dead previously? New code doesn't do this.

File: ee/clickhouse/util.py [hunk 6]
```diff
@@ -26,17 +27,28 @@ def assertNumQueries(self, num, func=None, *args, using=DEFAULT_DB_ALIAS, **kwar
 
     @contextmanager
     def capture_select_queries(self):
-        from ee.clickhouse.client import _annotate_tagged_query
+        queries = []
```
Comment (macobo) [inline ee/clickhouse/util.py:30]: Q: Why does this need a refactor?

Comment (hazzadous) [reply ee/clickhouse/util.py:30]: The specific thing I'm testing for is what is sent to clickhouse. Previously with capture_select_queries we were just looking at _annotate_tagged_query, which isn't the boundary to clickhouse.

I can revert I think, although it does seem a little brittle how it was before.

File: ee/clickhouse/client.py (hunk 5 above)
Comment (hazzadous) [reply ee/clickhouse/client.py:200]: `get_client()` appears to always return the same thing as far as I can tell. `ch_client` is variable. Although maybe I'm missing something 🤔

File: ee/clickhouse/client.py (hunk 3 above)
Comment (hazzadous) [reply ee/clickhouse/client.py:148]: Specifically this relates to https://github.com/mymarilyn/clickhouse-driver/blob/master/clickhouse_driver/client.py#L280

I can add a reference to this.

Comment (hazzadous) [reply ee/clickhouse/client.py:148]: It was quite confusing actually. The tests started to fail without this. There are some cases where we're catching all Exceptions like https://github.com/PostHog/posthog/blob/chore/remove-clickhouse-sql-comments/posthog/models/cohort.py#L192:L192

It took me some time to figure out what was going on. I could put in a separate PR to catch more specific errors, or maybe just change to reraise on `settings.TEST == True` as well?

File: ee/clickhouse/util.py (hunk 6 above)
Comment (macobo) [reply ee/clickhouse/util.py:30]: It's okay, was just surprised this needed a refactor.

File: ee/clickhouse/client.py [hunk 7]
```diff
@@ -130,40 +131,58 @@ def cache_sync_execute(query, args=None, redis_client=None, ttl=CACHE_TTL, setti
             redis_client.set(key, _serialize(result), ex=ttl)
             return result
 
+    InsertParams = Union[list, tuple, types.GeneratorType]
+    NonInsertParams = Union[Dict[str, Any]]
+    QueryArgs = Optional[Union[InsertParams, NonInsertParams]]
+
+    def prepare_query(client: SyncClient, query: str, args: QueryArgs):
```
Comment (macobo) [inline ee/clickhouse/client.py:138]: Nit: Let's order methods in order of importance here. `sync_execute` is the only function to be used outside of this file. Everything else should be _after_ that is possible, it's worth prepending a `_` to signal this is not "official" functionality either.

File: ee/clickhouse/client.py [hunk 8]
```diff
@@ -130,40 +131,58 @@ def cache_sync_execute(query, args=None, redis_client=None, ttl=CACHE_TTL, setti
             redis_client.set(key, _serialize(result), ex=ttl)
             return result
 
+    InsertParams = Union[list, tuple, types.GeneratorType]
+    NonInsertParams = Union[Dict[str, Any]]
+    QueryArgs = Optional[Union[InsertParams, NonInsertParams]]
```
Comment (macobo) [inline ee/clickhouse/client.py:136]: Nit: Type definitions to top of file - as-is they're placed kind of awkwardly.

**Review by macobo — APPROVED** [review]
Two nits but this looks good assuming tests will pass.

**Pull request merged**